src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -0,0 +1,120 @@
+        // uint add7Mask = (difference & 0x7070U) >> 4; // line YYY and ZZZ back up with the packed values
+        // uint packedResult = packedOriginalValues + add7Mask + 0x3030U /* ascii '0' */;
+
+        // The code below is equivalent to the commented out code above but has been tweaked
+        // to allow codegen to make some extra optimizations.
+
+        // The low byte of the packed result contains the hex representation of the incoming byte's low nibble.
+        // The adjacent byte of the packed result contains the hex representation of the incoming byte's high nibble.
+
+        // Finally, write to the output buffer starting with the *highest* index so that codegen can
+        // elide all but the first bounds check. (This only works if 'startingIndex' is a compile-time constant.)
+
+        // The JIT can elide bounds checks if 'startingIndex' is constant and if the caller is
+        // writing to a span of known length (or the caller has already checked the bounds of the
+        // furthest access).
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ToBytesBuffer(byte value, Span<byte> buffer, int startingIndex = 0, Casing casing = Casing.Upper)
+        {
+            uint difference = (((uint)value & 0xF0U) << 4) + ((uint)value & 0x0FU) - 0x8989U;
+            uint packedResult = ((((uint)(-(int)difference) & 0x7070U) >> 4) + difference + 0xB9B9U) | (uint)casing;
+
+            buffer[startingIndex + 1] = (byte)packedResult;
+            buffer[startingIndex] = (byte)(packedResult >> 8);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ToCharsBuffer(byte value, Span<char> buffer, int startingIndex = 0, Casing casing = Casing.Upper)
+        {
+            uint difference = (((uint)value & 0xF0U) << 4) + ((uint)value & 0x0FU) - 0x8989U;
+            uint packedResult = ((((uint)(-(int)difference) & 0x7070U) >> 4) + difference + 0xB9B9U) | (uint)casing;
+
+            buffer[startingIndex + 1] = (char)(packedResult & 0xFF);
+            buffer[startingIndex] = (char)(packedResult >> 8);
+        }
+
+        internal static unsafe string ToString(ReadOnlySpan<byte> bytes, Casing casing = Casing.Upper)
+        {
+#if NET45 || NET461 || NET462 || NET472 || NETSTANDARD1_0 || NETSTANDARD1_3 || NETSTANDARD2_0
+            Span<char> result = stackalloc char[bytes.Length * 2];
+            int pos = 0;
+            foreach (byte b in bytes)
+            {
+                ToCharsBuffer(b, result, pos, casing);
+                pos += 2;
+            }
+            return result.ToString();
+#else
+            fixed (byte* bytesPtr = bytes)
+            {
+                return string.Create(bytes.Length * 2, (Ptr: (IntPtr)bytesPtr, bytes.Length, casing), (chars, args) =>
+                {
+                    int pos = 0;
+                    foreach (byte b in new ReadOnlySpan<byte>((byte*)args.Ptr, args.Length))
```

Comment:
Why pay the cost of the iterator? A regular for loop would be cheaper.

Reply:
FWIW, the C# compiler actually translates it into the equivalent of a for loop rather than using its enumerator:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4foFcA7AM4BDAGYw62PhjoBZABRMAFkKgAqOgAcMUNBKl1cMPgHMMigJRsOXG5OkCeAWzoBeOgAYA3NZucR0GCEwRTo5YJU6MD06PhgAdzoAJUCAEwB5PlwATwBlDSE+AB4lFQA+OS0dAyNTC3MfX3ZKXxa6B2cAajcwb2bWgF8Gm0IAdjanXptByn6gA

Reply:
Ah ok. But the extra `pos` induction variable is still going to be there...

Reply:
I had it original as for loop but benchmark didn't show any improvements. Changed it back as it's perhaps more readable with for loop.